File: src/main/java/com/natu/ftax/ledger/infrastructure/LedgerEntryEntity.java
package com.natu.ftax.ledger.infrastructure;

import com.natu.ftax.ledger.domain.LedgerEntry;
import com.natu.ftax.transaction.domain.Token;
import jakarta.persistence.*;
import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;

import java.time.Instant;
import java.util.List;
import java.util.stream.Collectors;

@Entity
@Table(name = "ledger_entries")
public class LedgerEntryEntity {

    @Id
    private String id;

    private Instant instant;

    @ManyToOne(cascade = CascadeType.ALL)
    @JoinColumn(name = "ledger_book_id")
    @OnDelete(action = OnDeleteAction.CASCADE)
    private LedgerBookEntity ledgerBook;

    @ManyToMany(cascade = CascadeType.MERGE)
    @JoinTable(
        name = "ledger_entry_balance",
        joinColumns = @JoinColumn(name = "ledger_entry_id"),
        inverseJoinColumns = @JoinColumn(name = "balance_id")
    )
    @OnDelete(action = OnDeleteAction.CASCADE)
    private List<BalanceEntity> balances;

    protected LedgerEntryEntity() {
        // No-arg constructor required by JPA
    }

    public LedgerEntryEntity(String id, LedgerBookEntity ledgerBook,
            List<BalanceEntity> balances, Instant instant) {
        this.id = id;
        this.ledgerBook = ledgerBook;
        this.balances = balances;
        this.instant = instant;
    }

    public static LedgerEntryEntity fromDomain(LedgerEntry ledgerEntry, String ledgerBookId) {
        return new LedgerEntryEntity(
            ledgerEntry.getId(),
            new LedgerBookEntity(ledgerBookId),
            ledgerEntry.getBalances().values().stream().map(BalanceEntity::fromDomain).collect(Collectors.toList()),
                ledgerEntry.getInstant()
        );
    }

    public LedgerEntry toDomain() {
        return new LedgerEntry(
            id,
                balances.stream().collect(Collectors.toMap(e -> new Token(e.getToken()), BalanceEntity::toDomain)),
                instant
        );
    }
}

File: src/main/kotlin/com/natu/ftax/ledger/infrastructure/LedgerEntryEntity.kt
package com.natu.ftax.ledger.infrastructure


import com.natu.ftax.ledger.domain.LedgerEntry
import jakarta.persistence.*
import org.hibernate.annotations.OnDelete
import org.hibernate.annotations.OnDeleteAction
import java.util.Date

@Entity
@Table(name = "ledger_entries")
class LedgerEntryEntity(

    @Id
    val id: String,

    val date: Date,

    @ManyToOne(cascade = [CascadeType.ALL])
    @JoinColumn(name = "ledger_book_id")
    @OnDelete(action = OnDeleteAction.CASCADE)
    var ledgerBook: LedgerBookEntity,

    @ManyToMany(cascade = [CascadeType.MERGE])
    @JoinTable(
        name = "ledger_entry_balance",
        joinColumns = [JoinColumn(name = "ledger_entry_id")],
        inverseJoinColumns = [JoinColumn(name = "balance_id")]
    )
    @OnDelete(action = OnDeleteAction.CASCADE)
    val balances: MutableList<BalanceEntity>

) {

    // No-arg constructor required by JPA
    constructor() : this(id = "dum", ledgerBook = LedgerBookEntity("dum"), balances = mutableListOf(), date = Date())

    companion object {
        fun fromDomain(ledgerEntry: LedgerEntry, ledgerBookId: String): LedgerEntryEntity {
            return LedgerEntryEntity(
                id = ledgerEntry.id,
                ledgerBook = LedgerBookEntity(ledgerBookId),
                balances = ledgerEntry.balances.entries.map { BalanceEntity.fromDomain(it.value) }
                    .toMutableList(),
                date = ledgerEntry.date

            )
        }
    }

    fun toDomain(): LedgerEntry {
        val balancesMap = this.balances
            .map { it.toDomain() }
            .associateBy { it.token } // Creates a Map from the list
            .toMutableMap() // Converts the Map to a MutableMap
        return LedgerEntry(id, balancesMap, date)
    }
}